vg/recorder/recorder.go

```diff
@@ -0,0 +1,483 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package recorder provides support for vector graphics serialization
+// it is intended to be used as a testing helper.
+package recorder
+
+import (
+	"fmt"
+	"image/color"
+	"runtime"
+
+	"github.com/gonum/plot/vg"
+)
+
+var _ vg.Canvas = (*Canvas)(nil)
+
+// Canvas implements vg.Canvas operation serialization.
+type Canvas struct {
+	// Resolution holds the canvas resolution in DPI.
+	Resolution float64
+
+	// Actions holds a log of all methods called on
+	// the canvas.
+	Actions []Action
+
+	// KeepCaller indicates whether the Canvas will
+	// retain runtime caller location for the actions.
+	// This includes source filename and line number.
+	KeepCaller bool
+
+	// c holds a backing vg.Canvas. Method calls to
+	// the Canvas will be reflected to c.
+	c vg.Canvas
+
+	// fonts holds a collection of font/size descriptions.
+	fonts map[fontID]vg.Font
+}
+
+type fontID struct {
+	name string
+	size vg.Length
+}
+
+// Action is a vector graphics action as defined by the
+// vg.Canvas interface. Each method of vg.Canvas must have
+// a corresponding Action type.
+type Action interface {
+	Call() string
+	ApplyTo(vg.Canvas)
+	call() *location
+}
+
+type location struct {
+	haveCaller bool
+	file       string
+	line       int
+}
+
+func (l *location) setLocation() {
+	_, l.file, l.line, l.haveCaller = runtime.Caller(3)
+}
+
+func (l location) String() string {
+	if !l.haveCaller {
+		return ""
+	}
+	return fmt.Sprintf("%s:%d ", l.file, l.line)
+}
+
+// New returns a new Canvas with the specified resolution.
+func New(dpi float64) *Canvas {
+	return &Canvas{
+		Resolution: dpi,
+	}
+}
+
+// New returns a new Canvas from an existing vg.Canvas. vg.Canvas methods
+// called on the Canvas will also be passed to the provided backing
+// vg.Canvas.
+func NewFrom(c vg.Canvas) *Canvas {
+	return &Canvas{
+		Resolution: c.DPI(),
```

Comment:
Also note in the comment that the DPI of the returned canvas is inherited from c.